### PR TITLE
[GCE Discovery] Correcly handle large zones with 500 or more instances

### LINKED ISSUE #83783: Discovery GCE Pagination Lack
### Elasticsearch Version

All

### Installed Plugins

discovery-gce

### Java Version

_bundled_

### OS Version

any

### Problem Description

Currently, discovery-gce plugin tries to obtain all instances for a given project/zone in one shot.
Recently I've faced in my company (that had ~ 700 instances running at the time) a weird behavior of this plugin. The google api that this plugin uses paginates results in chunks of 500 items. The current implementation is ignoring if there's a next page available.

### Steps to Reproduce

- run the node discovery process in zones that have more than 500 instances

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by cdmbr**
discovery gce plugin has some issues while running in large pools of vms instances.

this pr attempts to solve it.

closes https://github.com/elastic/elasticsearch/issues/83783

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Thanks for the PR @cdmbr. Looks like you're still actively working on it so I've marked it as a draft for now. Please ping when it's ready for review. The following command will enable CI on your branch:

@elasticsearchmachine ok to test

**Comment by cdmbr** [comment]
thanks for the return @DaveCTurner !

I was just doing some lint fixes (forgot to run spotless 😅), but it seems that it is all set. I'm marking it as ready for review.

Question: when it merges, will it be back-ported for all previous versions ?

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

File: plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/nodeswithpagination/zones/europe-west1-b/instances?pageToken=next-token [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (arteam) [inline plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/nodeswithpagination/zones/europe-west1-b/instances?pageToken=next-token:1]: I believe question marks are not allowed in filenames on Windows, hence the Windows tests fail. 

What do you think about replacing the question mark with `%3F` from https://en.wikipedia.org/wiki/Percent-encoding and then make sure the question mark gets encoded in `GceMockUtils#readJsonResponse`?

```java
String mockFileName = Strings.replace(url, urlRoot, "")
            .replace("?", "%3F")
```

Comment (cdmbr) [reply plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/nodeswithpagination/zones/europe-west1-b/instances?pageToken=next-token]: thanks @arteam  ! fixed !

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java [hunk 2]
```diff
@@ -69,14 +70,20 @@
...
                 // hack around code messiness in GCE code
                 // TODO: get this fixed
-                InstanceList instanceList = Access.doPrivilegedIOException(() -> {
-                    Compute.Instances.List list = client().instances().list(project, zoneId);
-                    return list.execute();
+                return Access.doPrivilegedIOException(() -> {
+                    String nextPageToken = null;
+                    List<Instance> zoneInstances = new LinkedList<>();
+                    do {
+                        Compute.Instances.List list = client().instances().list(project, zoneId);
```
[the remark this answers is not in the document]
Comment (cdmbr) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: fixed !

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java [hunk 3]
```diff
@@ -69,14 +70,20 @@
             try {
                 // hack around code messiness in GCE code
                 // TODO: get this fixed
-                InstanceList instanceList = Access.doPrivilegedIOException(() -> {
-                    Compute.Instances.List list = client().instances().list(project, zoneId);
-                    return list.execute();
+                return Access.doPrivilegedIOException(() -> {
+                    String nextPageToken = null;
+                    List<Instance> zoneInstances = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (cdmbr) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: I used linkedlist here due to its O(1) in insertions ... But for traversals is not that good ...

File: plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceDiscoveryTests.java [hunk 4]
```diff
@@ -272,4 +272,15 @@ public void testMetadataServerValues() {
...
+
+    public void testNodesWithPagination() {
+        Settings nodeSettings = Settings.builder()
+            .put(GceInstancesServiceImpl.PROJECT_SETTING.getKey(), projectName)
+            .put(GceInstancesServiceImpl.ZONE_SETTING.getKey(), "europe-west1-b")
+            .putList(GceSeedHostsProvider.TAGS_SETTING.getKey(), "elasticsearch", "dev")
+            .build();
+        mock = new GceInstancesServiceMock(nodeSettings);
+        List<TransportAddress> dynamicHosts = buildDynamicNodes(mock, nodeSettings);
+        assertThat(dynamicHosts, hasSize(2));
```
Comment (fcofdez) [inline plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceDiscoveryTests.java:284]: We should assert that we get the expected `TransportAddress` values here.

**Comment by fcofdez** [comment]
@cdmbr Thanks for the contribution, it looks good, I've left a comment around the test that I think it should be addressed.

File: plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceDiscoveryTests.java (hunk 4 above)
Comment (cdmbr) [reply plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceDiscoveryTests.java:284]: no tests in this class does this assertion, only make sure that we have the expected number of distinct hosts ... do we really need to make this assertion here ?

Comment (fcofdez) [reply plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceDiscoveryTests.java:284]: It's to ensure that we get the expected results from the different responses.

**Review by arteam — APPROVED** [review]
LGTM! 

@cdmbr Thank you very much for working on this. I will wait for @fcofdez's response and then we should be good to merge this PR.

**Review by fcofdez — APPROVED** [review]
LGTM 👍 

Thanks for the PR @cdmbr!. @arteam would you take care of merging this and updating the changelog description?

**Comment by arteam** [comment]
@cdmbr Thank you very much for contributing to Elasticsearch!

**Pull request merged**